### PR TITLE
Cache civ uniques while refreshing world map tiles.

### PR DESCRIPTION
**Pull request by WhoIsJohannes**
This should have a pretty significant performance improvement of about 10% next round time.

Screenshot from 2023-04-25 19-26-44

### CODE REVIEW

**Comment by WhoIsJohannes** [comment]
Hm... I see yield icons flipping back and forth... 1 food one round, 2 food the other round and so on. What am I missing here? Nothing should change while the world map is refreshed right?

**Comment by WhoIsJohannes** [comment]
I found it... A lot of the code using the cache assumed that it is only reusing the cache across a single city, so it didn't matter that the key didn't include the city. But now I'm using the cache across several cities, so that doesn't work anymore. I fixed it now, but it feels ugly. I will check if there's a way to generate the cache key differently, so that this problem can't happen in the first place.

**Comment by WhoIsJohannes** [comment]
Done - I think this is ready for review now.

File: core/src/com/unciv/models/ruleset/unique/Unique.kt [hunk 1]
```diff
@@ -301,8 +301,23 @@ class LocalUniqueCache(val cache:Boolean = true) {
...
     private val keyToUniques = HashMap<String, Sequence<Unique>>()
 
+    fun forCityGetMatchingUniques(city: City, uniqueType: UniqueType) : Sequence<Unique> {
+        return forCityGetMatchingUniques(city, uniqueType, StateForConditionals(city.civ, city))
+    }
+    fun forCityGetMatchingUniques(city: City, uniqueType: UniqueType, stateForConditionals: StateForConditionals) : Sequence<Unique> {
+        return get("city-${city.id}-${uniqueType.name}-${stateForConditionals}", city.getMatchingUniques(uniqueType, stateForConditionals))
+    }
+
+    fun forCivGetMatchingUniques(civ: Civilization, uniqueType: UniqueType) : Sequence<Unique> {
```
Comment (yairm210) [inline core/src/com/unciv/models/ruleset/unique/Unique.kt:311]: state for conditionals should be an optional parameter defaulting to this and then we don't need 2 functions
Same for city functions

**Comment by yairm210** [comment]
Looks great! 👍🏿 
Love the generalizing of unique cache for city/civ!

**Comment by WhoIsJohannes** [comment]
Ah, good catch. For some reason I only had a boolean in mind. Do you mind
sending a PR? Thanks for investigating!

Il mer 26 apr 2023, 10:52 SomeTroglodyte ***@***.***> ha
scritto:

**Pull request merged**